lib/app_profiler/server.rb

```diff
@@ -0,0 +1,292 @@
+        end
+      end
+
+      class ProfileServerTransportTCP < ProfileServerTransport
+        SERVER_ADDRESS = "127.0.0.1" # it is ONLY safe to run this bound to localhost
+
+        def initialize(port = 0)
+          super()
+          FileUtils.mkdir_p(PROFILER_TEMPFILE_PATH)
+          @socket = TCPServer.new(SERVER_ADDRESS, port)
+          @port = @socket.addr[1]
+          @port_file = Tempfile.new("profileserver-#{Process.pid}-port-#{@port}-", PROFILER_TEMPFILE_PATH)
+        end
+
+        def client
+          TCPSocket.new(SERVER_ADDRESS, @port)
+        end
+
+        def stop
+          @port_file.unlink
+          @socket.close
+        end
+      end
+
+      def initialize(transport)
+        case transport
+        when TRANSPORT_UNIX
+          @transport = ProfileServer::ProfileServerTransportUNIX.new
+        when TRANSPORT_TCP
+          @transport = ProfileServer::ProfileServerTransportTCP.new(AppProfiler::Server.port)
+        else
+          raise "invalid transport #{transport}"
+        end
+
+        @listen_thread = nil
+
+        AppProfiler.logger.info(
+          "[AppProfiler::Server] listening on addr=#{@transport.socket.addr}"
+        )
+        at_exit { stop }
+      end
+
+      def client
+        @transport.client
+      end
+
+      def serve
+        return unless @listen_thread.nil?
+
+        app = ProfileApplication.new
+
+        @listen_thread = Thread.new do
+          loop do
+            Thread.new(@transport.socket.accept) do |session|
+              request = session.gets
+              if request.nil?
+                session.close
+                next
+              end
+              method, path, http_version = request.split(" ")
+              path_info, query_string = path.split("?")
+              env = { # an extremely minimal rack env hash, just enough to get the job done
+                "HTTP_VERSION" => http_version,
+                "REQUEST_METHOD" => method,
+                "PATH_INFO" => path_info,
+                "QUERY_STRING" => query_string,
+                "rack.input" => "",
+              }
+              status, headers, body = app.call(env)
+
+              begin
+                session.print("#{http_version} #{status}\r\n")
+                headers.each do |header, value|
+                  session.print("#{header}: #{value}\r\n")
+                end
+                session.print("\r\n")
+                body.each do |part|
+                  session.print(part)
+                end
+              rescue => e
+                AppProfiler.logger.error(
+                  "[AppProfiler::Server] exception #{e} responding to request #{request}: #{e.message}"
+                )
+              ensure
+                session.close
+              end
+            end
+          end
+        end
+      end
+
+      def stop
+        @listen_thread.kill
+        @transport.stop
+      end
+    end
+
+    class << self
+      def reset!
+        @profile_server = {}
```

Comment:
```suggestion
        profile_server.clear
```
might be more efficient.